In [6]:
import catboost
import sklearn
from sklearn.model_selection import train_test_split
from sklearn import ensemble

import pandas as pd
import numpy as np

In [7]:
pd.options.display.max_columns = None
df = pd.read_csv('../data/hp/train.csv', index_col='Id')
df

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1             60       RL         65.0     8450   Pave   NaN      Reg   
2             20       RL         80.0     9600   Pave   NaN      Reg   
3             60       RL         68.0    11250   Pave   NaN      IR1   
4             70       RL         60.0     9550   Pave   NaN      IR1   
5             60       RL         84.0    14260   Pave   NaN      IR1   
6             50       RL         85.0    14115   Pave   NaN      IR1   
7             20       RL         75.0    10084   Pave   NaN      Reg   
8             60       RL          NaN    10382   Pave   NaN      IR1   
9             50       RM         51.0     6120   Pave   NaN      Reg   
10           190       RL         50.0     7420   Pave   NaN      Reg   
11            20       RL         70.0    11200   Pave   NaN      Reg   
12            60       RL         85.0    11924   Pave   NaN      IR1   
13            20       RL          NaN    12968   Pave   NaN      IR2   
14            20       RL         91.0    10652   Pave   NaN      IR1   
15            20       RL          NaN    10920   Pave   NaN      IR1   
16            45       RM         51.0     6120   Pave   NaN      Reg   
17            20       RL          NaN    11241   Pave   NaN      IR1   
18            90       RL         72.0    10791   Pave   NaN      Reg   
19            20       RL         66.0    13695   Pave   NaN      Reg   
20            20       RL         70.0     7560   Pave   NaN      Reg   
21            60       RL        101.0    14215   Pave   NaN      IR1   
22            45       RM         57.0     7449   Pave  Grvl      Reg   
23            20       RL         75.0     9742   Pave   NaN      Reg   
24           120       RM         44.0     4224   Pave   NaN      Reg   
25            20       RL          NaN     8246   Pave   NaN      IR1   
26            20       RL        110.0    14230   Pave   NaN      Reg   
27            20       RL         60.0     7200   Pave   NaN      Reg   
28            20       RL         98.0    11478   Pave   NaN      Reg   
29            20       RL         47.0    16321   Pave   NaN      IR1   
30            30       RM         60.0     6324   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1431          60       RL         60.0    21930   Pave   NaN      IR3   
1432         120       RL          NaN     4928   Pave   NaN      IR1   
1433          30       RL         60.0    10800   Pave  Grvl      Reg   
1434          60       RL         93.0    10261   Pave   NaN      IR1   
1435          20       RL         80.0    17400   Pave   NaN      Reg   
1436          20       RL         80.0     8400   Pave   NaN      Reg   
1437          20       RL         60.0     9000   Pave   NaN      Reg   
1438          20       RL         96.0    12444   Pave   NaN      Reg   
1439          20       RM         90.0     7407   Pave   NaN      Reg   
1440          60       RL         80.0    11584   Pave   NaN      Reg   
1441          70       RL         79.0    11526   Pave   NaN      IR1   
1442         120       RM          NaN     4426   Pave   NaN      Reg   
1443          60       FV         85.0    11003   Pave   NaN      Reg   
1444          30       RL          NaN     8854   Pave   NaN      Reg   
1445          20       RL         63.0     8500   Pave   NaN      Reg   
1446          85       RL         70.0     8400   Pave   NaN      Reg   
1447          20       RL          NaN    26142   Pave   NaN      IR1   
1448          60       RL         80.0    10000   Pave   NaN      Reg   
1449          50       RL         70.0    11767   Pave   NaN      Reg   
1450         180       RM         21.0     1533   Pave   NaN      Reg   
1451          90       RL         60.0     9000   Pave   NaN      Reg   
1452          20       RL         78.0     9262   Pave   NaN    

- LotFrontage, LotArea, YearBuilt, YearRemodAdd, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF,
2ndFlrSF, LowQualFinSF, GrLivArea, TotRmsAbvGrd, Fireplaces, KitchenAbvGr, BedroomAbvGr,
FullBath, HalfBath, BsmtFullBath, BsmtHalfBath, GarageYrBlt, GarageCars, GarageArea, OverallQual, 
OverallCond,
WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, MoSold, YrSold - __numbers__

- Alley is weird 

- LotShape, LandContour, Utilities, LotConfig, LandSlope, Negihborhood, Condition1, Condition2, BldgType, HouseStyle, , RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure,
BsmtFinType1, BsmtFinType2, Heating, HeatingQC, CentralAir, Electrical, 
KitchenQual, Functional, FireplaceQu, GarageType, GarageFinish, GarageQual, GarageCond, 
PavedDrive, PoolQC, Fence, MiscFeature, SaleType, SaleCondition, Street, MSZoning, MSSubClass - __categoricals__

__SalePrice - the target variable (label)__

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 80 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non-

### now we must fill blanks in columns with less than 1460 values

In [8]:
df["LotFrontage"].fillna(value=0.0, inplace=True)
df["GarageYrBlt"].fillna(value=0.0, inplace=True)
df["MasVnrType"].fillna(value="0", inplace=True)
df["MasVnrArea"].fillna(value=0.0, inplace=True)
df["BsmtQual"].fillna(value="0", inplace=True)

df["BsmtCond"].fillna(value="0", inplace=True)
df["BsmtExposure"].fillna(value="0", inplace=True)
df["BsmtFinType1"].fillna(value="0", inplace=True)
df["BsmtFinType2"].fillna(value="0", inplace=True)
df["Electrical"].fillna(value="0", inplace=True)
df["FireplaceQu"].fillna(value="0", inplace=True)
df["GarageType"].fillna(value="0", inplace=True)
df["GarageFinish"].fillna(value="0", inplace=True)
df["GarageQual"].fillna(value="0", inplace=True)
df["GarageCond"].fillna(value="0", inplace=True)
df["Fence"].fillna(value="0", inplace=True)

In [9]:
# df["Alley"].fillna(value = '0', inplace = True)
# df["PoolQC"].fillna(value = '0', inplace = True)
del df["Alley"] 
del df["PoolQC"]
del df["MiscFeature"]
# delete PoolQC and MiscFeature because they are redundant, possibly delete alley because almost all is NaN

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 78 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1460 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            1460 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 no

In [10]:
CategoricalDataCols = ["LotShape", "LandContour", "Utilities", "LotConfig", "LandSlope", "Neighborhood", "Condition1", 
                       "Condition2", "BldgType", "HouseStyle", "RoofStyle", "RoofMatl", "Exterior1st", "Exterior2nd", 
                       "MasVnrType", "MasVnrArea", "ExterQual", "ExterCond", "Foundation", "BsmtQual", 
                       "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinType2", "Heating", "HeatingQC", 
                       "CentralAir", "Electrical", "KitchenQual", "Functional", "FireplaceQu", "GarageType", 
                       "GarageFinish", "GarageQual", "GarageCond", "PavedDrive", "Fence", "SaleType", 
                       "SaleCondition", "Street", "MSZoning", "MSSubClass"]

In [11]:
Labels = df["SalePrice"]
Data = df.drop("SalePrice", axis=1)


In [34]:
train_data

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                     
655          20       RL         91.0    10437   Pave     0      IR1   
233         160       RM         21.0     1680   Pave     0      Reg   
302          60       RL         66.0    16226   Pave     0      IR3   
1395        120       RL         53.0     4045   Pave     0      Reg   
1384         30       RL          0.0    25339   Pave     0      Reg   
1338         30       RM        153.0     4118   Pave  Grvl      IR1   
285         120       RL         50.0     8012   Pave     0      Reg   
1024        120       RL         43.0     3182   Pave     0      Reg   
738          60       RL         72.0    10463   Pave     0      IR1   
1002         30       RL         60.0     5400   Pave     0      Reg   
288          20       RL          0.0     8125   Pave     0      IR1   
198          75       RL        174.0    25419   Pave     0      Reg   
903          60       RL         63.0     7875   Pave     0      Reg   
47           50       RL         48.0    12822   Pave     0      IR1   
481          20       RL         98.0    16033   Pave     0      IR1   
318          60       FV         75.0     9000   Pave     0      Reg   
933          20       RL         84.0    11670   Pave     0      IR1   
199          75       RM         92.0     5520   Pave     0      Reg   
1075         20       RL         74.0     8556   Pave     0      Reg   
1019         80       RL          0.0    10784   Pave     0      IR1   
692          60       RL        104.0    21535   Pave     0      IR1   
871          20       RL         60.0     6600   Pave     0      Reg   
411          20       RL         68.0     9571   Pave     0      Reg   
1330         60       RL         63.0     9084   Pave     0      IR1   
32           20       RL          0.0     8544   Pave     0      IR1   
391          50       RL         50.0     8405   Pave  Grvl      Reg   
506          90       RM         60.0     7596   Pave  Grvl      Reg   
806          20       RL         91.0    12274   Pave     0      IR1   
16           45       RM         51.0     6120   Pave     0      Reg   
115          70       RL         61.0     7259   Pave     0      IR1   
...         ...      ...          ...      ...    ...   ...      ...   
800          50       RL         60.0     7200   Pave     0      Reg   
938          60       RL         75.0     9675   Pave     0      Reg   
1203         50       RM         50.0     6000   Pave     0      Reg   
762          30       RM         60.0     6911   Pave     0      Reg   
183          20       RL         60.0     9060   Pave     0      Reg   
724          50       RL         60.0     8172   Pave     0      Reg   
1382         20       RL          0.0    12925   Pave     0      IR1   
809          80       RL         85.0    13400   Pave     0      Reg   
319          60       RL         90.0     9900   Pave     0      Reg   
201          20       RM         80.0     8546   Pave     0      Reg   
43           85       RL          0.0     9180   Pave     0      IR1   
270          20       RL          0.0     7917   Pave     0      IR1   
111          50       RL         75.0     9525   Pave     0      Reg   
59           60       RL         66.0    13682   Pave     0      IR2   
863          20       RL         81.0     9672   Pave     0      Reg   
808          70       RL        144.0    21384   Pave     0      Reg   
1204         20       RL         75.0     9750   Pave     0      Reg   
335          60       RL         59.0     9042   Pave     0      IR1   
815          50       RL         45.0     8248   Pave  Grvl      Reg   
694          30       RL         60.0     5400   Pave     0      Reg   
742          20       RL         65.0     6768   Pave     0      IR1   
188          50       RL         60.0    10410   Pave     0      Reg   
250          50       RL          0.0   15900

Now let's one-hot encode all the categoricals

In [12]:
for col in CategoricalDataCols:
    Data[col] = pd.Categorical(Data[col])

train_data, test_data, train_labels, test_labels = train_test_split(Data, Labels, random_state=98987, test_size=0.25)

In [13]:
train_data_cats = train_data[CategoricalDataCols]
test_data_cats = test_data[CategoricalDataCols]
oh_train_data = train_data
for col in CategoricalDataCols:
    oh_train_data = oh_train_data.drop(col, axis=1)
oh_test_data = test_data
for col in CategoricalDataCols:
    oh_test_data = oh_test_data.drop(col, axis=1)

display(oh_test_data.head())

LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
Id                                                                              
883           0.0     9636            6            5       1992          1993   
1443         85.0    11003           10            5       2008          2008   
1136         60.0     6180            6            5       1926          1950   
631          50.0     9000            5            6       1880          1991   
1067         59.0     7837            6            7       1993          1994   

      BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  2ndFlrSF  \
Id                                                                         
883            0           0        808          808       808       785   
1443         765           0        252         1017      1026       981   
1136           0           0        960          960       986         0   
631            0           0        636          636      1089       661   
1067           0           0        799          799       799       772   

      LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
Id                                                                              
883              0       1593             0             0         2         1   
1443             0       2007             1             0         2         1   
1136             0        986             0             0         1         0   
631              0       1750             0             0         1         0   
1067             0       1571             0             0         2         1   

      BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  \
Id                                                                        
883              3             1             7           1       1993.0   
1443             3             1            10           1       2008.0   
1136             2             1             5           1       1926.0   
631              3             1             8           0       1937.0   
1067             3             1             7           1       1993.0   

      GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
Id                                                                     
883            2         389         342           40              0   
1443           3         812         168           52              0   
1136           1         180           0          128              0   
631            1         240           0            0            293   
1067           2         380           0           40              0   

      3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  
Id                                                               
883           0            0         0        0      12    2009  
1443          0            0         0        0       4    2009  
1136          0            0         0        0       5    2007  
631           0            0         0        0       6    2006  
1067          0            0         0        0       5    2009

In [58]:
display(train_data_cats.head())

LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
Id                                                                     
655       IR1         Lvl    AllPub    Inside       Gtl      NoRidge   
233       Reg         Lvl    AllPub    Inside       Gtl       BrDale   
302       IR3         Lvl    AllPub    Inside       Gtl      CollgCr   
1395      Reg         Lvl    AllPub    Inside       Gtl      Blmngtn   
1384      Reg         Lvl    AllPub    Inside       Gtl       Sawyer   

     Condition1 Condition2 BldgType HouseStyle RoofStyle RoofMatl Exterior1st  \
Id                                                                              
655        Norm       Norm     1Fam     1Story       Hip  CompShg     MetalSd   
233        Norm       Norm    Twnhs     2Story     Gable  CompShg     HdBoard   
302        Norm       Norm     1Fam     2Story     Gable  CompShg     VinylSd   
1395       Norm       Norm   TwnhsE     1Story       Hip  CompShg     VinylSd   
1384       Norm       Norm     1Fam     1Story     Gable  CompShg     Wd Sdng   

     Exterior2nd MasVnrType MasVnrArea ExterQual ExterCond Foundation  \
Id                                                                      
655      MetalSd    BrkFace      660.0        Gd        Gd      PConc   
233      HdBoard    BrkFace      297.0        TA        TA     CBlock   
302      VinylSd       None        0.0        Gd        TA      PConc   
1395     VinylSd    BrkFace       45.0        Gd        TA      PConc   
1384     Wd Sdng       None        0.0        TA        Gd     BrkTil   

     BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2 Heating  \
Id                                                                      
655        Gd       TA           Gd          GLQ          Unf    GasA   
233        TA       TA           No          Unf          Unf    GasA   
302        Gd       TA           No          GLQ          Unf    GasA   
1395       Gd       TA           Av          GLQ          Unf    GasA   
1384       TA       TA           No          Unf          Unf    GasA   

     HeatingQC CentralAir Electrical KitchenQual Functional FireplaceQu  \
Id                                                                        
655         Ex          Y      SBrkr          Gd        Typ          TA   
233         TA          Y      SBrkr          TA        Typ          Po   
302         Ex          Y      SBrkr          Gd        Typ          TA   
1395        Ex          Y      SBrkr          Gd        Typ          Gd   
1384        Ex          Y      SBrkr          Gd        Typ           0   

     GarageType GarageFinish GarageQual GarageCond PavedDrive Fence SaleType  \
Id                                                                             
655      Attchd          Fin         TA         TA          Y     0       WD   
233      Attchd          Unf         TA         TA          Y     0       WD   
302     BuiltIn          Fin         TA         TA          Y     0       WD   
1395     Attchd          Fin         TA         TA          Y     0      New   
1384     Attchd          Unf         TA         TA          N     0       WD   

     SaleCondition Street MSZoning MSSubClass  
Id                                             
655         Normal   Pave       RL         20  
233         Normal   Pave       RM        160  
302         Normal   Pave       RL         60  
1395       Partial   Pave       RL        120  
1384        Normal   Pave       RL         30

In [57]:
display(test_data_cats.head())

LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
Id                                                                     
883       IR1         Lvl    AllPub    Corner       Gtl      Gilbert   
1443      Reg         Lvl    AllPub    Inside       Gtl      Somerst   
1136      Reg         Lvl    AllPub    Corner       Gtl      BrkSide   
631       Reg         Lvl    AllPub    Corner       Gtl      OldTown   
1067      IR1         Lvl    AllPub    Inside       Gtl      Gilbert   

     Condition1 Condition2 BldgType HouseStyle RoofStyle RoofMatl Exterior1st  \
Id                                                                              
883        Norm       Norm     1Fam     2Story     Gable  CompShg     VinylSd   
1443       Norm       Norm     1Fam     2Story     Gable  CompShg     VinylSd   
1136       Norm       Norm     1Fam     1Story     Gable  CompShg     Wd Sdng   
631      Artery       Norm     1Fam     2Story     Gable  CompShg     VinylSd   
1067       Norm       Norm     1Fam     2Story     Gable  CompShg     VinylSd   

     Exterior2nd MasVnrType MasVnrArea ExterQual ExterCond Foundation  \
Id                                                                      
883      VinylSd       None        0.0        TA        TA      PConc   
1443     VinylSd      Stone      160.0        Ex        TA      PConc   
1136     Wd Sdng       None        0.0        TA        TA     BrkTil   
631      VinylSd       None        0.0        TA        TA     BrkTil   
1067     VinylSd       None        0.0        Gd        TA      PConc   

     BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2 Heating  \
Id                                                                      
883        Gd       TA           No          Unf          Unf    GasA   
1443       Ex       TA           Av          GLQ          Unf    GasA   
1136       TA       TA           No          Unf          Unf    GasA   
631        Fa       Fa           No          Unf          Unf    GasA   
1067       Gd       TA           No          Unf          Unf    GasA   

     HeatingQC CentralAir Electrical KitchenQual Functional FireplaceQu  \
Id                                                                        
883         Gd          Y      SBrkr          TA        Typ          TA   
1443        Ex          Y      SBrkr          Ex        Typ          Ex   
1136        TA          N      SBrkr          TA        Typ          Gd   
631         TA          Y      FuseA          Ex        Typ           0   
1067        Gd          Y      SBrkr          TA        Typ          TA   

     GarageType GarageFinish GarageQual GarageCond PavedDrive  Fence SaleType  \
Id                                                                              
883     BuiltIn          RFn         TA         TA          Y  MnPrv       WD   
1443     Attchd          Fin         TA         TA          Y      0       WD   
1136     Detchd          Unf         TA         TA          Y      0       WD   
631      Detchd          Unf         Fa         Po          N  MnPrv       WD   
1067     Attchd          RFn         TA         TA          Y      0       WD   

     SaleCondition Street MSZoning MSSubClass  
Id                                             
883         Normal   Pave       RL         60  
1443        Normal   Pave       FV         60  
1136        Normal   Pave       RM         30  
631        Abnorml   Pave       RM         70  
1067        Normal   Pave       RL         60

In [14]:
one_hot = sklearn.preprocessing.OneHotEncoder(handle_unknown='ignore')
train_data_cats = one_hot.fit_transform(train_data_cats).toarray()

In [15]:
test_data_cats = one_hot.transform(test_data_cats).toarray()

In [16]:
for i in range(len(train_data_cats[0])):
#     print(len(train_data_cats))
#     print(len(train_data_cats[0]))
#     print
#     print(i)
    oh_train_data['one-hot' + str(i)] = train_data_cats[:, i]
    
for i in range(len(test_data_cats[0])):
    oh_test_data['one-hot' + str(i)] = test_data_cats[:, i]

display(oh_train_data.head())
display(oh_train_data.head())

LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
Id                                                                              
655          91.0    10437            8            6       1995          1995   
233          21.0     1680            6            5       1972          1972   
302          66.0    16226            8            5       1998          1999   
1395         53.0     4045            7            5       2006          2006   
1384          0.0    25339            5            7       1918          2007   

      BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  2ndFlrSF  \
Id                                                                         
655         1696           0        413         2109      2113         0   
233            0           0        483          483       483       504   
302          281           0        747         1028      1210      1242   
1395        1070           0        286         1356      1500         0   
1384           0           0        816          816      1416         0   

      LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
Id                                                                              
655              0       2113             1             0         2         1   
233              0        987             0             0         1         1   
302              0       2452             0             0         2         1   
1395             0       1500             1             0         2         0   
1384             0       1416             0             0         2         0   

      BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  \
Id                                                                        
655              2             1             7           1       1995.0   
233              2             1             5           1       1972.0   
302              4             1             9           1       1998.0   
1395             2             1             6           1       2006.0   
1384             3             1             7           0       2007.0   

      GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
Id                                                                     
655            3         839         236           46              0   
233            1         288           0            0              0   
302            2         683         208           50              0   
1395           3         648         161           20              0   
1384           2         576           0            0            112   

      3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  one-hot0  \
Id                                                                          
655           0            0         0        0       8    2008       1.0   
233           0            0         0        0       6    2006       0.0   
302           0            0         0        0       5    2007       0.0   
1395          0            0         0        0      10    2006       0.0   
1384          0            0         0        0       8    2007       0.0   

      one-hot1  one-hot2  one-hot3  one-hot4  one-hot5  one-hot6  one-hot7  \
Id                                                                           
655        0.0       0.0       0.0       0.0       0.0       0.0       1.0   
233        0.0       0.0       1.0       0.0       0.0       0.0       1.0   
302        0.0       1.0       0.0       0.0       0.0       0.0       1.0   
1395       0.0       0.0       1.0       0.0       0.0       0.0       1.0   
1384       0.0       0.0       1.0       0.0       0.0       0.0       1.0   

      one-hot8  one-hot9  one-hot10  one-hot11  one-hot12  one-hot13  \
Id                                                                     
655        1.0       0.0        0.0        0.0        0.0        1.0   
233

LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
Id                                                                              
655          91.0    10437            8            6       1995          1995   
233          21.0     1680            6            5       1972          1972   
302          66.0    16226            8            5       1998          1999   
1395         53.0     4045            7            5       2006          2006   
1384          0.0    25339            5            7       1918          2007   

      BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  2ndFlrSF  \
Id                                                                         
655         1696           0        413         2109      2113         0   
233            0           0        483          483       483       504   
302          281           0        747         1028      1210      1242   
1395        1070           0        286         1356      1500         0   
1384           0           0        816          816      1416         0   

      LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
Id                                                                              
655              0       2113             1             0         2         1   
233              0        987             0             0         1         1   
302              0       2452             0             0         2         1   
1395             0       1500             1             0         2         0   
1384             0       1416             0             0         2         0   

      BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  \
Id                                                                        
655              2             1             7           1       1995.0   
233              2             1             5           1       1972.0   
302              4             1             9           1       1998.0   
1395             2             1             6           1       2006.0   
1384             3             1             7           0       2007.0   

      GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
Id                                                                     
655            3         839         236           46              0   
233            1         288           0            0              0   
302            2         683         208           50              0   
1395           3         648         161           20              0   
1384           2         576           0            0            112   

      3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  one-hot0  \
Id                                                                          
655           0            0         0        0       8    2008       1.0   
233           0            0         0        0       6    2006       0.0   
302           0            0         0        0       5    2007       0.0   
1395          0            0         0        0      10    2006       0.0   
1384          0            0         0        0       8    2007       0.0   

      one-hot1  one-hot2  one-hot3  one-hot4  one-hot5  one-hot6  one-hot7  \
Id                                                                           
655        0.0       0.0       0.0       0.0       0.0       0.0       1.0   
233        0.0       0.0       1.0       0.0       0.0       0.0       1.0   
302        0.0       1.0       0.0       0.0       0.0       0.0       1.0   
1395       0.0       0.0       1.0       0.0       0.0       0.0       1.0   
1384       0.0       0.0       1.0       0.0       0.0       0.0       1.0   

      one-hot8  one-hot9  one-hot10  one-hot11  one-hot12  one-hot13  \
Id                                                                     
655        1.0       0.0        0.0        0.0        0.0        1.0   
233

In [81]:
sklearn_boost = ensemble.GradientBoostingRegressor(loss='ls', learning_rate=0.1, n_estimators=100, 
                                   subsample=1.0, criterion='friedman_mse', min_samples_split=2, 
                                   min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, 
                                   min_impurity_decrease=0.0, min_impurity_split=None, init=None, 
                                   random_state=98987, max_features=None, alpha=0.9, verbose=0, 
                                   max_leaf_nodes=None, warm_start=False, presort='deprecated', 
                                   validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)

sklearn_boost.fit(oh_train_data, train_labels)
print("train score", sklearn_boost.score(oh_train_data, train_labels))
print("test score", sklearn_boost.score(oh_test_data, test_labels))

train score 0.9725603672039187
test score 0.8420232319410241


In [83]:
cboost = catboost.CatBoostRegressor(iterations=None,
                        learning_rate=None,
                        depth=None,
                        l2_leaf_reg=None,
                        model_size_reg=None,
                        rsm=None,
                        loss_function='RMSE',
                        border_count=None,
                        feature_border_type=None,
                        per_float_feature_quantization=None,
                        input_borders=None,
                        output_borders=None,
                        fold_permutation_block=None,
                        od_pval=None,
                        od_wait=None,
                        od_type=None,
                        nan_mode=None,
                        counter_calc_method=None,
                        leaf_estimation_iterations=None,
                        leaf_estimation_method=None,
                        thread_count=None,
                        random_seed=None,
                        use_best_model=None,
                        best_model_min_trees=None,
                        verbose=False,
                        silent=None,
                        logging_level=None,
                        metric_period=None,
                        ctr_leaf_count_limit=None,
                        store_all_simple_ctr=None,
                        max_ctr_complexity=None,
                        has_time=None,
                        allow_const_label=None,
                        one_hot_max_size=None,
                        random_strength=None,
                        name=None,
                        ignored_features=None,
                        train_dir=None,
                        custom_metric=None,
                        eval_metric=None,
                        bagging_temperature=None,
                        save_snapshot=None,
                        snapshot_file=None,
                        snapshot_interval=None,
                        fold_len_multiplier=None,
                        used_ram_limit=None,
                        gpu_ram_part=None,
                        pinned_memory_size=None,
                        allow_writing_files=None,
                        final_ctr_computation_mode=None,
                        approx_on_full_history=None,
                        boosting_type=None,
                        simple_ctr=None,
                        combinations_ctr=None,
                        per_feature_ctr=None,
                        ctr_target_border_count=None,
                        task_type=None,
                        device_config=None,                        
                        devices=None,
                        bootstrap_type=None,
                        subsample=None,                        
                        sampling_unit=None,
                        dev_score_calc_obj_block_size=None,
                        max_depth=None,
                        n_estimators=None,
                        num_boost_round=None,
                        num_trees=None,
                        colsample_bylevel=None,
                        random_state=None,
                        reg_lambda=None,
                        objective=None,
                        eta=None,
                        max_bin=None,
                        gpu_cat_features_storage=None,
                        data_partition=None,
                        metadata=None,
                        early_stopping_rounds=None,
                        cat_features=None,
                        grow_policy=None,
                        min_data_in_leaf=None,
                        min_child_samples=None,
                        max_leaves=None,
                        num_leaves=None,
                        score_function=None,
                        leaf_estimation_backtracking=None,
                        ctr_history_unit=None,
                        monotone_constraints=None)
cboost.fit(oh_train_data, train_labels)
print("test score", cboost.score(oh_test_data, test_labels))
print("train score", cboost.score(oh_train_data, train_labels))

test score 0.8472655334171177
train score 0.9895222095429762


In [89]:
sklearn_boost = ensemble.GradientBoostingRegressor(loss='ls', learning_rate=0.1, n_estimators=100, 
                                   subsample=1.0, criterion='friedman_mse', min_samples_split=2, 
                                   min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, 
                                   min_impurity_decrease=0.0, min_impurity_split=None, init=None, 
                                   random_state=98987, max_features=None, alpha=0.9, verbose=0, 
                                   max_leaf_nodes=None, warm_start=False, presort='deprecated', 
                                   validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)


params = {
          'max_depth': [3, 5 ,10, 2],
          'max_features': [None, 'sqrt'],

          }

GradBoostGrid = sklearn.model_selection.GridSearchCV(sklearn_boost, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)
GradBoostGrid.fit(oh_train_data, train_labels)

display(pd.DataFrame(GradBoostGrid.cv_results_).sort_values('rank_test_score').head())

print("test score", GradBoostGrid.score(oh_test_data, test_labels))
print("train score", GradBoostGrid.score(oh_train_data, train_labels))

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
3       0.082211      0.000797         0.001994    1.266988e-06   
1       0.057471      0.000484         0.001996    1.818245e-06   
0       0.756243      0.007716         0.002001    1.191750e-05   
2       1.191184      0.005688         0.001998    5.631122e-06   
7       0.046077      0.000399         0.001995    1.907349e-07   

  param_max_depth param_max_features  \
3               5               sqrt   
1               3               sqrt   
0               3               None   
2               5               None   
7               2               sqrt   

                                     params  split0_test_score  \
3  {'max_depth': 5, 'max_features': 'sqrt'}           0.859416   
1  {'max_depth': 3, 'max_features': 'sqrt'}           0.869393   
0    {'max_depth': 3, 'max_features': None}           0.886059   
2    {'max_depth': 5, 'max_features': None}           0.871143   
7  {'max_depth': 2, 'max_features': 'sqrt'}           0.855777   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
3           0.881565           0.905737           0.753564           0.933711   
1           0.891446           0.890312           0.738320           0.930508   
0           0.905024           0.891479           0.696342           0.933899   
2           0.897761           0.890684           0.721687           0.916377   
7           0.890400           0.882699           0.742561           0.913935   

   mean_test_score  std_test_score  rank_test_score  
3         0.866799        0.061787                1  
1         0.863996        0.065869                2  
0         0.862561        0.084746                3  
2         0.859530        0.070426                4  
7         0.857075        0.060192                5

test score 0.8087165180595882
train score 0.9825962641169265


#### now i will normalize data

In [90]:
minmax = sklearn.preprocessing.MinMaxScaler()
oh_train_data = minmax.fit_transform(oh_train_data)
oh_test_data = minmax.transform(oh_test_data)

In [92]:
display(oh_test_data)

array([[0.        , 0.0510284 , 0.55555556, ..., 0.        , 0.        ,
        0.        ],
       [0.2715655 , 0.05939643, 1.        , ..., 0.        , 0.        ,
        0.        ],
       [0.19169329, 0.02987267, 0.55555556, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.33546326, 0.04389079, 0.22222222, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.19076273, 0.33333333, ..., 0.        , 0.        ,
        1.        ],
       [0.23961661, 0.05402179, 0.55555556, ..., 0.        , 0.        ,
        0.        ]])

In [93]:
sklearn_boost = ensemble.GradientBoostingRegressor(loss='ls', learning_rate=0.1, n_estimators=100, 
                                   subsample=1.0, criterion='friedman_mse', min_samples_split=2, 
                                   min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, 
                                   min_impurity_decrease=0.0, min_impurity_split=None, init=None, 
                                   random_state=98987, max_features=None, alpha=0.9, verbose=0, 
                                   max_leaf_nodes=None, warm_start=False, presort='deprecated', 
                                   validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)

sklearn_boost.fit(oh_train_data, train_labels)
print("train score", sklearn_boost.score(oh_train_data, train_labels))
print("test score", sklearn_boost.score(oh_test_data, test_labels))

train score 0.9725603672039187
test score 0.8419843008677319


In [94]:
cboost = catboost.CatBoostRegressor(iterations=None,
                        learning_rate=None,
                        depth=None,
                        l2_leaf_reg=None,
                        model_size_reg=None,
                        rsm=None,
                        loss_function='RMSE',
                        border_count=None,
                        feature_border_type=None,
                        per_float_feature_quantization=None,
                        input_borders=None,
                        output_borders=None,
                        fold_permutation_block=None,
                        od_pval=None,
                        od_wait=None,
                        od_type=None,
                        nan_mode=None,
                        counter_calc_method=None,
                        leaf_estimation_iterations=None,
                        leaf_estimation_method=None,
                        thread_count=None,
                        random_seed=None,
                        use_best_model=None,
                        best_model_min_trees=None,
                        verbose=False,
                        silent=None,
                        logging_level=None,
                        metric_period=None,
                        ctr_leaf_count_limit=None,
                        store_all_simple_ctr=None,
                        max_ctr_complexity=None,
                        has_time=None,
                        allow_const_label=None,
                        one_hot_max_size=None,
                        random_strength=None,
                        name=None,
                        ignored_features=None,
                        train_dir=None,
                        custom_metric=None,
                        eval_metric=None,
                        bagging_temperature=None,
                        save_snapshot=None,
                        snapshot_file=None,
                        snapshot_interval=None,
                        fold_len_multiplier=None,
                        used_ram_limit=None,
                        gpu_ram_part=None,
                        pinned_memory_size=None,
                        allow_writing_files=None,
                        final_ctr_computation_mode=None,
                        approx_on_full_history=None,
                        boosting_type=None,
                        simple_ctr=None,
                        combinations_ctr=None,
                        per_feature_ctr=None,
                        ctr_target_border_count=None,
                        task_type=None,
                        device_config=None,                        
                        devices=None,
                        bootstrap_type=None,
                        subsample=None,                        
                        sampling_unit=None,
                        dev_score_calc_obj_block_size=None,
                        max_depth=None,
                        n_estimators=None,
                        num_boost_round=None,
                        num_trees=None,
                        colsample_bylevel=None,
                        random_state=None,
                        reg_lambda=None,
                        objective=None,
                        eta=None,
                        max_bin=None,
                        gpu_cat_features_storage=None,
                        data_partition=None,
                        metadata=None,
                        early_stopping_rounds=None,
                        cat_features=None,
                        grow_policy=None,
                        min_data_in_leaf=None,
                        min_child_samples=None,
                        max_leaves=None,
                        num_leaves=None,
                        score_function=None,
                        leaf_estimation_backtracking=None,
                        ctr_history_unit=None,
                        monotone_constraints=None)
cboost.fit(oh_train_data, train_labels)
print("test score", cboost.score(oh_test_data, test_labels))
print("train score", cboost.score(oh_train_data, train_labels))

test score 0.8472753135866607
train score 0.9895222095429762


In [95]:
sklearn_boost = ensemble.GradientBoostingRegressor(loss='ls', learning_rate=0.1, n_estimators=100, 
                                   subsample=1.0, criterion='friedman_mse', min_samples_split=2, 
                                   min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, 
                                   min_impurity_decrease=0.0, min_impurity_split=None, init=None, 
                                   random_state=98987, max_features=None, alpha=0.9, verbose=0, 
                                   max_leaf_nodes=None, warm_start=False, presort='deprecated', 
                                   validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)


params = {
          'max_depth': [3, 5 ,10, 2],
          'max_features': [None, 'sqrt'],

          }

GradBoostGrid = sklearn.model_selection.GridSearchCV(sklearn_boost, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)
GradBoostGrid.fit(oh_train_data, train_labels)

display(pd.DataFrame(GradBoostGrid.cv_results_).sort_values('rank_test_score').head())

print("test score", GradBoostGrid.score(oh_test_data, test_labels))
print("train score", GradBoostGrid.score(oh_train_data, train_labels))

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
3       0.091898      0.001214         0.001402        0.000483   
1       0.062463      0.000815         0.001795        0.000399   
0       0.909186      0.018882         0.001197        0.000399   
2       1.392431      0.007597         0.001388        0.000496   
7       0.049455      0.001339         0.001003        0.000013   

  param_max_depth param_max_features  \
3               5               sqrt   
1               3               sqrt   
0               3               None   
2               5               None   
7               2               sqrt   

                                     params  split0_test_score  \
3  {'max_depth': 5, 'max_features': 'sqrt'}           0.859287   
1  {'max_depth': 3, 'max_features': 'sqrt'}           0.869403   
0    {'max_depth': 3, 'max_features': None}           0.885818   
2    {'max_depth': 5, 'max_features': None}           0.871131   
7  {'max_depth': 2, 'max_features': 'sqrt'}           0.855804   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
3           0.881422           0.905797           0.753730           0.933624   
1           0.891471           0.890312           0.738320           0.930613   
0           0.905024           0.891312           0.696342           0.933899   
2           0.897803           0.890544           0.721912           0.916437   
7           0.890492           0.882699           0.742588           0.913935   

   mean_test_score  std_test_score  rank_test_score  
3         0.866772        0.061711                1  
1         0.864024        0.065892                2  
0         0.862479        0.084721                3  
2         0.859565        0.070340                4  
7         0.857104        0.060192                5

test score 0.8083198428149783
train score 0.9825962641169265


__there should be xgboost and lightgbm here, but later.__
#### now with PCA

In [97]:
from sklearn.decomposition import PCA
pcator = PCA(n_components=0.9, copy=True, whiten=False, svd_solver='full', tol=0.0, iterated_power='auto', random_state=None)
## MUST USE FIT_TRANSFORM, will break otherwise because of copy=false
oh_train_data_reduced = pcator.fit_transform(oh_train_data)
print("done fitting")
oh_test_data_reduced = pcator.transform(oh_test_data)

done fitting


In [98]:
cboost = catboost.CatBoostRegressor(iterations=None,
                        learning_rate=None,
                        depth=None,
                        l2_leaf_reg=None,
                        model_size_reg=None,
                        rsm=None,
                        loss_function='RMSE',
                        border_count=None,
                        feature_border_type=None,
                        per_float_feature_quantization=None,
                        input_borders=None,
                        output_borders=None,
                        fold_permutation_block=None,
                        od_pval=None,
                        od_wait=None,
                        od_type=None,
                        nan_mode=None,
                        counter_calc_method=None,
                        leaf_estimation_iterations=None,
                        leaf_estimation_method=None,
                        thread_count=None,
                        random_seed=None,
                        use_best_model=None,
                        best_model_min_trees=None,
                        verbose=False,
                        silent=None,
                        logging_level=None,
                        metric_period=None,
                        ctr_leaf_count_limit=None,
                        store_all_simple_ctr=None,
                        max_ctr_complexity=None,
                        has_time=None,
                        allow_const_label=None,
                        one_hot_max_size=None,
                        random_strength=None,
                        name=None,
                        ignored_features=None,
                        train_dir=None,
                        custom_metric=None,
                        eval_metric=None,
                        bagging_temperature=None,
                        save_snapshot=None,
                        snapshot_file=None,
                        snapshot_interval=None,
                        fold_len_multiplier=None,
                        used_ram_limit=None,
                        gpu_ram_part=None,
                        pinned_memory_size=None,
                        allow_writing_files=None,
                        final_ctr_computation_mode=None,
                        approx_on_full_history=None,
                        boosting_type=None,
                        simple_ctr=None,
                        combinations_ctr=None,
                        per_feature_ctr=None,
                        ctr_target_border_count=None,
                        task_type=None,
                        device_config=None,                        
                        devices=None,
                        bootstrap_type=None,
                        subsample=None,                        
                        sampling_unit=None,
                        dev_score_calc_obj_block_size=None,
                        max_depth=None,
                        n_estimators=None,
                        num_boost_round=None,
                        num_trees=None,
                        colsample_bylevel=None,
                        random_state=None,
                        reg_lambda=None,
                        objective=None,
                        eta=None,
                        max_bin=None,
                        gpu_cat_features_storage=None,
                        data_partition=None,
                        metadata=None,
                        early_stopping_rounds=None,
                        cat_features=None,
                        grow_policy=None,
                        min_data_in_leaf=None,
                        min_child_samples=None,
                        max_leaves=None,
                        num_leaves=None,
                        score_function=None,
                        leaf_estimation_backtracking=None,
                        ctr_history_unit=None,
                        monotone_constraints=None)
cboost.fit(oh_train_data_reduced, train_labels)
print("test score", cboost.score(oh_test_data_reduced, test_labels))
print("train score", cboost.score(oh_train_data_reduced, train_labels))

test score 0.7474930922954502
train score 0.9954828054465529


In [108]:
sklearn_boost = ensemble.GradientBoostingRegressor(loss='ls', learning_rate=0.1, n_estimators=100, 
                                   subsample=1.0, criterion='friedman_mse', min_samples_split=2, 
                                   min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, 
                                   min_impurity_decrease=0.0, min_impurity_split=None, init=None, 
                                   random_state=98987, max_features=None, alpha=0.9, verbose=0, 
                                   max_leaf_nodes=None, warm_start=False, presort='deprecated', 
                                   validation_fraction=0.1, n_iter_no_change=None, tol=0.0001, ccp_alpha=0.0)


params = {
          'max_depth': [3, 5, 2, 7],
          'max_features': [None, 'sqrt', 'log2'],

          }

GradBoostGrid = sklearn.model_selection.GridSearchCV(sklearn_boost, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)
GradBoostGrid.fit(oh_train_data_reduced, train_labels)

display(pd.DataFrame(GradBoostGrid.cv_results_).sort_values('rank_test_score').head())

print("test score", GradBoostGrid.score(oh_test_data_reduced, test_labels))
print("train score", GradBoostGrid.score(oh_train_data_reduced, train_labels))

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
3       2.122795      0.004948         0.000997        0.000003   
0       1.359100      0.011802         0.000998        0.000002   
4       0.260013      0.002833         0.001596        0.000489   
1       0.167567      0.001810         0.001004        0.000012   
5       0.181717      0.001593         0.001197        0.000399   

  param_max_depth param_max_features  \
3               5               None   
0               3               None   
4               5               sqrt   
1               3               sqrt   
5               5               log2   

                                     params  split0_test_score  \
3    {'max_depth': 5, 'max_features': None}           0.730183   
0    {'max_depth': 3, 'max_features': None}           0.733401   
4  {'max_depth': 5, 'max_features': 'sqrt'}           0.680057   
1  {'max_depth': 3, 'max_features': 'sqrt'}           0.665928   
5  {'max_depth': 5, 'max_features': 'log2'}           0.664853   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
3           0.717999           0.814010           0.622721           0.806964   
0           0.729272           0.791528           0.628052           0.804840   
4           0.685422           0.767564           0.714332           0.794398   
1           0.721603           0.769828           0.664166           0.799086   
5           0.694187           0.743121           0.688031           0.799366   

   mean_test_score  std_test_score  rank_test_score  
3         0.738376        0.069690                1  
0         0.737419        0.062477                2  
4         0.728355        0.045324                3  
1         0.724122        0.054215                4  
5         0.717912        0.048044                5

test score 0.6853514520595103
train score 0.9962506901672723


#### PCA makes everything significantly worse

In [25]:
import xgboost as xgb

In [113]:
xgboost = xgb.XGBRegressor()
xgboost.fit(oh_train_data, train_labels)
xgboost.score(oh_test_data, test_labels)

d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \


[16:37:01] WARNING: C:/Jenkins/workspace/xgboost-win64_release_0.90/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


0.845356133220309

In [2]:
import lightgbm as lgb

In [18]:
lgbreg = lgb.LGBMRegressor(boosting_type='gbdt', num_leaves=31, max_depth=-1, 
                           learning_rate=0.1, n_estimators=100, subsample_for_bin=200000,
                           objective=None, class_weight=None, min_split_gain=0.0, 
                           min_child_weight=0.001, min_child_samples=20, subsample=1.0, 
                           subsample_freq=0, colsample_bytree=1.0, reg_alpha=0.0, 
                           reg_lambda=0.0, random_state=None, n_jobs=-1, silent=True, importance_type='split')

lgbreg.fit(oh_train_data, train_labels)
print(lgbreg.score(oh_test_data, test_labels))
print(lgbreg.score(oh_train_data, train_labels))

0.8299393997016506
0.9811004059839733


In [24]:
params = {
    'boosting_type': ['gbdt'],
    'max_depth': [-1, 1, 3, 5, 7],
    'learning_rate': [0.1, 0.05, 0.5, 1.],
    'n_estimators': [20, 50, 100],
}


lgbgrid = sklearn.model_selection.GridSearchCV(lgbreg, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)

lgbgrid.fit(oh_train_data, train_labels)

display(pd.DataFrame(lgbgrid.cv_results_).sort_values('rank_test_score').head())
print('train score', lgbgrid.score(oh_train_data, train_labels))
print('test score', lgbgrid.score(oh_test_data, test_labels))

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
8        0.157976      0.003817         0.107715        0.001260   
11       0.177727      0.005688         0.110907        0.001932   
17       0.261706      0.006807         0.109707        0.002092   
1        0.197673      0.004649         0.111108        0.001850   
29       0.224003      0.006203         0.115292        0.001620   

   param_boosting_type param_learning_rate param_max_depth param_n_estimators  \
8                 gbdt                 0.1               3                100   
11                gbdt                 0.1               5                100   
17                gbdt                0.05              -1                100   
1                 gbdt                 0.1              -1                 50   
29                gbdt                0.05               7                100   

                                               params  split0_test_score  \
8   {'boosting_type': 'gbdt', 'learning_rate': 0.1...           0.863156   
11  {'boosting_type': 'gbdt', 'learning_rate': 0.1...           0.865608   
17  {'boosting_type': 'gbdt', 'learning_rate': 0.0...           0.848908   
1   {'boosting_type': 'gbdt', 'learning_rate': 0.1...           0.849375   
29  {'boosting_type': 'gbdt', 'learning_rate': 0.0...           0.847582   

    split1_test_score  split2_test_score  split3_test_score  \
8            0.917473           0.882947           0.777665   
11           0.911858           0.874206           0.780643   
17           0.912025           0.876280           0.794514   
1            0.913861           0.873721           0.793797   
29           0.912490           0.876316           0.795709   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
8            0.935346         0.875317        0.054996                1  
11           0.932810         0.873025        0.052294                2  
17           0.925262         0.871398        0.046889                3  
1            0.926106         0.871372        0.047557                4  
29           0.922497         0.870919        0.046084                5

train score 0.9527467409844161
test score 0.8285823618167415


In [28]:
xgboost = xgb.XGBRegressor(max_depth=3, learning_rate=0.1, n_estimators=100, 
                            verbosity=1, objective='reg:squarederror', 
                            booster='gbtree', tree_method='auto', n_jobs=1, 
                            gamma=0, min_child_weight=1, max_delta_step=0, 
                            subsample=1, colsample_bytree=1, colsample_bylevel=1, 
                            colsample_bynode=1, reg_alpha=0, reg_lambda=1, 
                            scale_pos_weight=1, base_score=0.5, 
                            random_state=1998, missing=None, num_parallel_tree=1, 
                            importance_type='gain')


params = {
    'max_depth': [1, 3, 5, 10, 20],
    'learning_rate': [0.1, 0.001, 0.5], 
    'n_estimators': [100, 50]
}


xgrid = sklearn.model_selection.GridSearchCV(xgboost, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)

xgrid.fit(oh_train_data, train_labels)

display(pd.DataFrame(xgrid.cv_results_).sort_values('rank_test_score').head())
print('train score', xgrid.score(oh_train_data, train_labels))
print('test score', xgrid.score(oh_test_data, test_labels))

d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
d:\utilites\programming\anaconda-python\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future vers

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
4       2.074148      0.036912         0.009791        0.001166   
5       1.046822      0.022761         0.008784        0.000747   
2       1.420881      0.015894         0.008977        0.000631   
7       1.726460      0.013507         0.008981        0.000003   
9       2.413032      0.028376         0.008977        0.000003   

  param_learning_rate param_max_depth param_n_estimators  \
4                 0.1               5                100   
5                 0.1               5                 50   
2                 0.1               3                100   
7                 0.1              10                 50   
9                 0.1              20                 50   

                                              params  split0_test_score  \
4  {'learning_rate': 0.1, 'max_depth': 5, 'n_esti...           0.880136   
5  {'learning_rate': 0.1, 'max_depth': 5, 'n_esti...           0.874474   
2  {'learning_rate': 0.1, 'max_depth': 3, 'n_esti...           0.872850   
7  {'learning_rate': 0.1, 'max_depth': 10, 'n_est...           0.856949   
9  {'learning_rate': 0.1, 'max_depth': 20, 'n_est...           0.851541   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
4           0.889462           0.883297           0.752627           0.923885   
5           0.883075           0.879905           0.757372           0.916890   
2           0.897860           0.881375           0.696956           0.931173   
7           0.878537           0.868324           0.731488           0.913991   
9           0.873631           0.875883           0.731429           0.911873   

   mean_test_score  std_test_score  rank_test_score  
4         0.865881        0.058743                1  
5         0.862343        0.054552                2  
2         0.856043        0.082007                3  
7         0.849858        0.062190                4  
9         0.848871        0.061825                5

train score 0.9904536178943112
test score 0.850203133852062


In [29]:
cboost = catboost.CatBoostRegressor(iterations=None,
                        learning_rate=None,
                        depth=None,
                        l2_leaf_reg=None,
                        model_size_reg=None,
                        rsm=None,
                        loss_function='RMSE',
                        border_count=None,
                        feature_border_type=None,
                        per_float_feature_quantization=None,
                        input_borders=None,
                        output_borders=None,
                        fold_permutation_block=None,
                        od_pval=None,
                        od_wait=None,
                        od_type=None,
                        nan_mode=None,
                        counter_calc_method=None,
                        leaf_estimation_iterations=None,
                        leaf_estimation_method=None,
                        thread_count=None,
                        random_seed=98987,
                        use_best_model=None,
                        best_model_min_trees=None,
                        verbose=False,
                        silent=None,
                        logging_level=None,
                        metric_period=None,
                        ctr_leaf_count_limit=None,
                        store_all_simple_ctr=None,
                        max_ctr_complexity=None,
                        has_time=None,
                        allow_const_label=None,
                        one_hot_max_size=None,
                        random_strength=None,
                        name=None,
                        ignored_features=None,
                        train_dir=None,
                        custom_metric=None,
                        eval_metric=None,
                        bagging_temperature=None,
                        save_snapshot=None,
                        snapshot_file=None,
                        snapshot_interval=None,
                        fold_len_multiplier=None,
                        used_ram_limit=None,
                        gpu_ram_part=None,
                        pinned_memory_size=None,
                        allow_writing_files=None,
                        final_ctr_computation_mode=None,
                        approx_on_full_history=None,
                        boosting_type=None,
                        simple_ctr=None,
                        combinations_ctr=None,
                        per_feature_ctr=None,
                        ctr_target_border_count=None,
                        task_type=None,
                        device_config=None,                        
                        devices=None,
                        bootstrap_type=None,
                        subsample=None,                        
                        sampling_unit=None,
                        dev_score_calc_obj_block_size=None,
                        max_depth=None,
                        n_estimators=None,
                        num_boost_round=None,
                        num_trees=None,
                        colsample_bylevel=None,
                        random_state=None,
                        reg_lambda=None,
                        objective=None,
                        eta=None,
                        max_bin=None,
                        gpu_cat_features_storage=None,
                        data_partition=None,
                        metadata=None,
                        early_stopping_rounds=None,
                        cat_features=None,
                        grow_policy=None,
                        min_data_in_leaf=None,
                        min_child_samples=None,
                        max_leaves=None,
                        num_leaves=None,
                        score_function=None,
                        leaf_estimation_backtracking=None,
                        ctr_history_unit=None,
                        monotone_constraints=None)


params = {
    'depth': [1, 3, 5, 10, 20],
    'learning_rate': [0.01, 0.1, 0.5, 1.]
    
    
}


catgrid = sklearn.model_selection.GridSearchCV(cboost, params, scoring=None, n_jobs=None, 
                                                iid='deprecated', refit=True, cv=None, verbose=0, 
                                                pre_dispatch='2*n_jobs', return_train_score=False)

catgrid.fit(oh_train_data, train_labels)

display(pd.DataFrame(catgrid.cv_results_).sort_values('rank_test_score').head())
print('train score', catgrid.score(oh_train_data, train_labels))
print('test score', catgrid.score(oh_test_data, test_labels))

d:\utilites\programming\anaconda-python\lib\site-packages\sklearn\model_selection\_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
_catboost.CatBoostError: c:/goagent/pipelines/buildmaster/catboost.git/catboost/private/libs/options/oblivious_tree_options.cpp:117: Maximum tree depth is 16

  FitFailedWarning)


mean_fit_time  std_fit_time  mean_score_time  std_score_time param_depth  \
9        4.318367      0.578589         0.024541        0.003265           5   
5        2.072917      0.161696         0.033910        0.019493           3   
8        4.235502      0.639911         0.027726        0.004820           5   
13      30.910411      0.329409         0.034718        0.020054          10   
12      29.881345      0.394771         0.024942        0.000888          10   

   param_learning_rate                                params  \
9                  0.1    {'depth': 5, 'learning_rate': 0.1}   
5                  0.1    {'depth': 3, 'learning_rate': 0.1}   
8                 0.01   {'depth': 5, 'learning_rate': 0.01}   
13                 0.1   {'depth': 10, 'learning_rate': 0.1}   
12                0.01  {'depth': 10, 'learning_rate': 0.01}   

    split0_test_score  split1_test_score  split2_test_score  \
9            0.897353           0.910334           0.898616   
5            0.904240           0.917998           0.897883   
8            0.869650           0.907819           0.898802   
13           0.836177           0.863800           0.910513   
12           0.828901           0.865790           0.904451   

    split3_test_score  split4_test_score  mean_test_score  std_test_score  \
9            0.751617           0.942763         0.880137        0.066316   
5            0.712303           0.947789         0.876042        0.083655   
8            0.770883           0.932913         0.876013        0.056323   
13           0.808993           0.923192         0.868535        0.043277   
12           0.818219           0.920924         0.867657        0.040351   

    rank_test_score  
9                 1  
5                 2  
8                 3  
13                4  
12                5

train score 0.9976869549915889
test score 0.8494437893941841
